.NET for Apache Spark: ML Sentiment Analysis (#17344)
dotnet/docs: docs/spark/tutorials/ml-sentment-analysis.md

---
title: Sentiment analysis with .NET for Apache Spark and ML.NET tutorial
description: In this tutorial, you learn how to use ML.NET with .NET for Apache Spark for sentiment analysis.
author: mamccrea
ms.author: mamccrea
ms.date: 03/07/2019
ms.topic: tutorial
---

# Tutorial: Sentiment analysis with .NET for Apache Spark and ML.NET

> This tutorial teaches you how to perform sentiment analysis of online reviews using ML.NET and .NET for Apache Spark. ML.NET is a free, cross-platform, open-source machine learning framework. You can use ML.NET with .NET for Apache Spark to scale the training and prediction of machine learning algorithms.

Reviewer: Do we want to hyperlink to http://dot.net/ml on `ML.NET`?

In this tutorial, you learn how to:

> [!div class="checklist"]
>
> * Write a .NET for Apache Spark app that performs sentiment analysis.

> > * Use the ML.NET model builder in Visual Studio.

Reviewer: ```suggestion
> * Use the ML.NET Model Builder in Visual Studio.
```

> * 

## Prerequisites

If this is your first .NET for Apache Spark application, start with the [Getting Started tutorial](get-started.md) to become familiar with the basics.

This tutorial uses the ML.NET Model Builder (preview), a visual interface available in Visual Studio. If you do not already have Visual Studio, you can [download the Community version of Visual Studio](https://visualstudio.microsoft.com/downloads/) for free.

[Download and install](https://marketplace.visualstudio.com/items?itemName=MLNET.07) ML.NET Model Builder (preview).

> 

Reviewer: Should we provide any sort of docs links to learn more about Sentiment Analysis and/or ML.NET?

## Build your machine learning model

1. Open Visual Studio and create a new C# Console App (.NET Core). Name the project *MLSparkModel*.

1. In **Solution Explorer**, right-click the *MLSparkModel* project. Then select **Add > Machine Learning**.

1. From the ML.NET Model Builder, select the **Sentiment Analysis** scenario tile.

> 1. On the **Add data** page, upload the *Yelp* reviews data set. 

Reviewer: Add a link/explanation for how to download the Yelp dataset, and maybe some context about how it contains online Yelp reviews about various services?

Reviewer: Also, do we want to explain the data within Yelp.csv, i.e. one column of review text, and another column with 1 or 0 for positive or negative sentiment?

1. Search for **Microsoft.Spark** and install the package. **Microsoft.ML** was installed for you by Model Builder.

### Create a SparkSession

1. Open the *Program.cs* file for **MLSparkModelML.ConsoleApp**. This file was auto-generated by Model Builder. Delete the `using` statements, the contents of the Main() method, and section for sample data generation.

1. Add the following additional `using` statements to the top of the *Program.cs*:

   ```csharp
   using System;
   using System.Collections.Generic;
   using Microsoft.ML;
   using Microsoft.ML.Data;
   using Microsoft.Spark.Sql;
   ```

1. Add the following code to your `Main` method to create a new `SparkSession`. The Spark Session is the entry point to programming Spark with the Dataset and DataFrame API.

   ```csharp
   SparkSession spark = SparkSession
        .Builder()
        .AppName(".NET for Apache Spark Sentiment Analysis")
        .GetOrCreate();
   ```

   Calling the *spark* object created above allows you to access Spark and DataFrame functionality throughout your program.

### Create a DataFrame and print to console

> Read in the Yelp review data from the *yelp.csv* file as a `DataFrame`. Include `header` and `inferSchema` options.

Reviewer: Maybe we could add some info around why we include `header` and `inferSchema`:

```suggestion
Read in the Yelp review data from the *yelp.csv* file as a `DataFrame`. Include `header` and `inferSchema` options. This will indicate that *yelp.csv* has a header row with column labels, and it will automatically infer the data's schema, which is the name and type of data in each column.
```

Reviewer: Also potentially add a statement about replacing `DATA_FILEPATH` with the path to yelp.csv on your machine?

```csharp
DataFrame df = spark
    .ReadStream()
    .Option("header", true)
    .Option("inferSchema", true)
    .Csv(DATA_FILEPATH);

df.Show();
```

### Register a user-defined function

> You can use UDFs, *user-defined functions*, in Spark applications to perform calculations and analysis on your data. In this how-to, you use ML.NET with a UDF to evaluate each review.

Reviewer: ```suggestion
You can use UDFs, *user-defined functions*, in Spark applications to perform calculations and analysis on your data. In this tutorial, you use ML.NET with a UDF to evaluate each Yelp review.
```

Add the following code to your `Main` method to register a UDF called `MLudf`. 

```csharp
spark.Udf()
    .Register<ModelInput, ModelOutput>("MLudf", (text) => ConsumeModel.Predict(text));
```

This UDF takes a Yelp review string as input, and outputs true or false for positive and negative sentiments, respectively. It uses the *ConsumeModel()* method that was created by Model Builder.

### Use Spark SQL to call the UDF

Now that you've read in your data and incorporated ML, use Spark SQL to call the UDF that will run sentiment analysis on each row of your DataFrame. Add the following code to your `Main` method:

```csharp
// Use Spark SQL to call ML.NET UDF
// Display results of sentiment analysis on reviews
df.CreateOrReplaceTempView("Reviews");

> DataFrame sqlDf = spark.Sql("SELECT ReviewText, MLudf(ReviewText) FRReviews");

Reviewer: ```suggestion
DataFrame sqlDf = spark.Sql("SELECT ReviewText, MLudf(ReviewText) FROM Reviews");
```

### Create a SparkSession

1. Open the *Program.cs* file for **MLSparkModelML.ConsoleApp**. This file was autogenerated by Model Builder. Delete the `using` statements, the contents of the Main() method, and the `CreateSingleDataSample` region.

1. Add the following additional `using` statements to the top of the *Program.cs*:

   ```csharp
   using System;
   using System.Collections.Generic;
   using Microsoft.ML;
   using Microsoft.ML.Data;
   using Microsoft.Spark.Sql;
   using MLSparkModelML.Model;
   ```

1. Change the `DATA_FILEPATH` to the path of your *yelptest.csv*.

1. Add the following code to your `Main` method to create a new `SparkSession`. The Spark Session is the entry point to programming Spark with the Dataset and DataFrame API.

   ```csharp
   SparkSession spark = SparkSession
        .Builder()
        .AppName(".NET for Apache Spark Sentiment Analysis")
        .GetOrCreate();
   ```

   Calling the *spark* object created above allows you to access Spark and DataFrame functionality throughout your program.

### Create a DataFrame and print to console

> Read in the Yelp review data from the *yelptest.csv* file as a `DataFrame`. Include `header` and `inferSchema` options.

Reviewer: Do we want to explain why we're using `header` and `inferSchema` or what they do?

```csharp
DataFrame df = spark
    .ReadStream()
    .Option("header", true)
    .Option("inferSchema", true)
    .Csv(DATA_FILEPATH);

df.Show();
```

### Register a user-defined function

You can use UDFs, *user-defined functions*, in Spark applications to do calculations and analysis on your data. In this tutorial, you use ML.NET with a UDF to evaluate each Yelp review.

Add the following code to your `Main` method to register a UDF called `MLudf`.

```csharp
spark.Udf()

>     .Register<string, bool>("MLudf", predict);

Reviewer: I think this UDF would need some more info/to call an actual `predict()` method?

In my sample, the UDF is defined as:

`
spark.Udf().Register<string, bool>(
                "MLudf",
                (text) => Sentiment(text, args[1]));
`

Author: The text is passed in when the UDF is called in the Spark SQL statement

Reviewer: Yes I see that. But the Spark SQL statement takes in a string `ReviewText` as input- where does this definition of MLudf actually use a string? I see it's defined to take in a string parameter, but where does it actually define and use a string?

Author: First, we define predict(text). Then we register the UDF, which only requires the name, predict. Then we use MLudf(text) in Spark SQL, which passes in a string. 

Your function uses Sentiment(text, modelpath), but only the text is passed in the Spark SQL statement. I think that's why it needs to be added during UDF registration.

Reviewer: Interesting, I didn't know that was possible! I thought `predict` would at least require parentheses to indicate it's a method. 

Thanks for the detailed explanation, and now I learned something new :)

```

> This UDF takes a Yelp review string as input, and outputs true or false for positive and negative sentiments, respectively. It uses the *predict()* method that you define in a later step.

Reviewer: ```suggestion
This UDF takes a Yelp review string as input, and outputs true or false for positive or negative sentiments, respectively. It uses the *predict()* method that you define in a later step.
```

### Use Spark SQL to call the UDF

Now that you've read in your data and incorporated ML, use Spark SQL to call the UDF that will run sentiment analysis on each row of your DataFrame. Add the following code to your `Main` method:

```csharp
// Use Spark SQL to call ML.NET UDF
// Display results of sentiment analysis on reviews
df.CreateOrReplaceTempView("Reviews");
DataFrame sqlDf = spark.Sql("SELECT ReviewText, MLudf(ReviewText) FROM Reviews");
sqlDf.Show();

// Print out first 20 rows of data
// Prevent data getting cut off by setting truncate = 0
sqlDf.Show(20, 0, false);

spark.Stop();
```

### Create predict() method

> Add the following code before you `Main()` method. This code is similar to what is produced by Model Builder in *ConsumeModel.cs*. Moving this method to your console keeps the model loading every time you run your app.

Reviewer: ```suggestion
Add the following code before your `Main()` method. This code is similar to what is produced by Model Builder in *ConsumeModel.cs*. Moving this method to your console keeps the model loading every time you run your app.
```

Reviewer: Is there a reason we should have users add it before `Main()`? I would think it would be okay if it was before or after `Main`.

> Add the following code before your `Main()` method. This code is similar to what is produced by Model Builder in *ConsumeModel.cs*. Moving this method to your console keeps the model loading every time you run your app.

Reviewer 2: The last sentence took me a couple reads to understand. Would this be simpler:

```suggestion
Add the following code before your `Main()` method. This code is similar to what is produced by Model Builder in *ConsumeModel.cs*. Moving this method to your console loads the model  each time you run your app.
```

```csharp
private static readonly PredictionEngine<ModelInput, ModelOutput> _predictionEngine;

> static Program()

Reviewer 2: This took me a few times through to figure out that you were using a static constructor. I'd recommend adding a comment on that.

More importantly, it's very difficult to catch exceptions thrown in static constructors. That makes errors in them very hard to debug. For that reason, we usually recommend that static ctors are simple, with minimal logic. In this example, loading the zip file for the model could fail for a variety of reasons, like copying it to the wrong folder. Would that make it hard for readers to diagnose and fix? 

I'd consider refactoring this when you have time.

    _predictionEngine = mlContext.Model.CreatePredictionEngine<ModelInput, ModelOutput>(model);
}

static bool predict(string text)
{
    ModelInput input = new ModelInput
    {
        ReviewText = text
    };

    return _predictionEngine.Predict(input).Prediction;
}
```

## Add the model to your console app

In Solution Explorer, copy the *MLModel.zip* file from the **MLSparkModelML.Model** project and paste it in the **MLSparkModelML.ConsoleApp** project. A reference is automatically added in *MLSparkModelML.ConsoleApp.csproj*.

## Run your code

Use `spark-submit` to run your code. Navigate to your console app's root folder using the command prompt and run the following commands.

First, clean and publish your app.

```dotnetcli
dotnet clean
dotnet publish
```

Then navigate to the console app's publish folder and run the following `spark-submit` command. Remember to update the command with the actual path of your Microsoft Spark jar file.

```dotnetcli
%SPARK_HOME%\bin\spark-submit --class org.apache.spark.deploy.dotnet.DotnetRunner --master local microsoft-spark-2.4.x-0.10.0.jar dotnet MLSparkModelML.ConsoleApp.dll
```

## Get the code

> This tutorial uses code from the [Sentiment Analysis with Big Data](https://github.com/dotnet/spark/tree/master/examples/Microsoft.Spark.CSharp.Examples/MachineLearning/Sentiment) example.

Reviewer 2: I looked at the code in this folder, and it doesn't match the code in the article. Was this adapting an earlier sample? Or is the code to match in another PR?

Author: No, I adapted this code based on Model Builder output. I changed the wording to note that it's similar - not identical.
